src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,956 @@
+                // `null or (not L1, _, ...) or (_, not L2, ...) or ...`
+
+                var builder = ArrayBuilder<BoundPattern>.GetInstance();
+
+                if (node.InputType.CanContainNull())
+                {
+                    // `null`
+                    builder.Add(new BoundConstantPattern(node.Syntax,
+                        new BoundLiteral(node.Syntax, constantValueOpt: null, type: null, hasErrors: false), // dummy expression
+                        ConstantValue.Null, node.InputType, node.NarrowedType, hasErrors: false).MakeCompilerGenerated());
+                }
+
+                var subpatterns = node.Subpatterns;
+                var discards = subpatterns.SelectAsArray(d => d.WithPattern(MakeDiscardPattern(d)));
+
+                for (int i = 0; i < subpatterns.Length; i++)
+                {
+                    var negatedPattern = (BoundPattern)Visit(subpatterns[i].Pattern);
+
+                    if (IsNegatedDiscard(negatedPattern))
+                    {
+                        continue;
+                    }
+
+                    // `(..., not LN, ...)`
+                    builder.Add(node.WithSubpatterns(discards.SetItem(i, subpatterns[i].WithPattern(negatedPattern))).WithSyntax(subpatterns[i].Syntax));
+                }
+
+                BoundPattern result = MakeDisjunction(node, builder);
+                builder.Free();
+                return result;
+            }
+
+            public override BoundNode? VisitListPattern(BoundListPattern node)
+            {
+                if (!_negated)
+                {
+                    return base.VisitListPattern(node);
+                }
+
+                // `[L1, L2, ...]`
+                // can be expanded to `not null and [_, _, ...] and [L1, _, ...] and [_, L2, ...] and ...`
+                // which can be negated to
+                // `null or not [_, _, ...] or [not L1, _, ...] or [_, not L2, ...] or ...`
+
+                var builder = ArrayBuilder<BoundPattern>.GetInstance();
+
+                // TODO2 not for structs
+                if (node.InputType.CanContainNull())
+                {
+                    // `null`
+                    builder.Add(new BoundConstantPattern(node.Syntax,
+                        new BoundLiteral(node.Syntax, constantValueOpt: null, type: null, hasErrors: false), // dummy expression
+                        ConstantValue.Null, node.InputType, node.NarrowedType, hasErrors: false).MakeCompilerGenerated());
+                }
+
+                ImmutableArray<BoundPattern> discardSubpatterns = node.Subpatterns.SelectAsArray(replaceWithDiscards);
+
+                // `not [_, _, ..., .._]`
+                BoundListPattern listOfDiscards = node.WithSubpatterns(discardSubpatterns);
+                if (!ListPatternHasOnlyEmptySlice(listOfDiscards))
+                {
+                    builder.Add(new BoundNegatedPattern(node.Syntax,
+                        listOfDiscards,
```

Comment:
>listOfDiscards [](http://example.com/codeflow?start=24&length=14)

The name of this local is extremely confusing. It took me a very very long time to realize that, if slice is present, this is actually not a list of discards and what we are doing in that case actually makes sense.